### ISSUE TITLE
StaticHtmlRenderer escapes apostrophes/quotes unnecessarily, breaking script blocks

### ISSUE THREAD

**Issue by oliverw** [dotnet/aspnetcore#52547]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Might be related to https://github.com/dotnet/aspnetcore/issues/49797.

Component:

```
<script>notifier.info('@Message');</script>

@code {
    [Parameter]
    public string Message { get; set; }
```

Used as:

`<InfoToast Message="yo"></InfoToast>`

still renders as:

`<script>notifier.info(\u0027yo\u0027);</script>`

### .NET Version

8.0.100

**Comment by danroth27** [comment]
Repro: https://github.com/danroth27/RazorHtmlRendering/tree/issue52547

```html
<script>notifier.info(\u0027yo\u0027);</script>
```

**Comment by danroth27** [comment]
@oliverw We agree that this is a bug, but we have limited resources, so we have to triage issues based on their severity and user impact. Our initial assessment of this issue is that it is specific to static rendering when using C# to dynamically render JavaScript, which we don't consider to be mainline scenario. If you think we've underestimated the severity of this issue, please let us know your reasoning. It helps if you can share specific details about what you are trying to accomplish, so that we can understand better how broadly impacting this issue is. There may also be an alternative approach that doesn't run into this issue.

Lastly, this is a community-based project, so if this issue is particularly important to your scenario, we do accept contributions. Just send us a PR and we'll consider it. Any contribution from the community to help make ASP.NET Core & Blazor better for everyone is much appreciated!

**Comment by MackinnonBuck** [comment]
If we fix this issue, we would probably need to do so in the Razor compiler. There are two relevant `RenderTreeBuilder` APIs here that the Razor compiler generates calls to:

* `AddMarkupContent`
* `AddContent`

The `AddMarkupContent` API does not HTML-encode content, but the `AddContent` API does. When a C# value gets rendered in Razor markup, the generated call is `AddContent`, because we don't want the value to somehow impact the DOM structure of the page. However, content inside a `script` really shouldn't be HTML encoded because the browser's HTML parser knows that the content inside `<script>` is "script content". So, part of the fix would be for the Razor compiler to replace `AddContent` calls with `AddMarkupContent` when inside a script element.

However, that's not completely what we're seeing here. Somehow, adding the C# variable to the script causes _all_ content inside the script element to be HTML-encoded, including the quotes before and after the C# variable. Removing the `@` results in a well-formed script. Regardless of whether we should be encouraging C# values to be injected into `<script>` elements, this definitely seems like a bug.

There's also strangely a difference between...

```razor
<!-- Does not work -->
<script>console.log('@Message');</script>
```

...and:

```razor
<!-- Works -->
<script>
	console.log('@Message');
</script>
```

...which means we can't rely on this bug as a way to discourage customers from injecting computed values into scripts.

Without being an expert in the Razor compiler's internals or the logic that creates this bug, my suggestion would be:
* Update the behavior so content inside a `<script>` element always uses the `AddMarkupContent` API.
* Add guidance to suggest being extremely cautious when injecting computed values inside `<script>` elements.
  * We could even go further and emit a diagnostic warning when doing this.

**Comment by SteveSandersonMS** [comment]
> Update the behavior so content inside a <script> element always uses the AddMarkupContent API.

I think we'd be expected to use JavaScript string encoding by default, otherwise it's an XSS vector by default. People should explicitly cast to markup strings if they want the output to be unencoded.

**Comment by mkArtakMSFT** [comment]
> Regardless of whether we should be encouraging C# values to be injected into `<script>` elements, this definitely seems like a bug.
> 
> There's also strangely a difference between...
> 
> ```
> 
> <script>console.log('@Message');</script>
> ```
> 
> ...and:
> 
> ```
> 
> <script>
> 	console.log('@Message');
> </script>
> ```
> 
> ...which means we can't rely on this bug as a way to discourage customers from injecting computed values into scripts.

@jjonescz this seems to still be a bug. even though it seems you tried to fix it earlier: https://github.com/dotnet/razor/issues/9204 ?


Do you know what may be causing this behavior change ?

**Comment by jjonescz** [comment]
> this seems to still be a bug. even though it seems you tried to fix it earlier: [dotnet/razor#9204](https://github.com/dotnet/razor/issues/9204) ?

That previous bug was about `<script>` tags being special-cased - the fix was to remove that special casing. This bug is different, it happens for `<script>` tags, `<div>` tags, etc.

> Do you know what may be causing this behavior change ?

This is caused by an optimization inside the razor compiler. If I understand correctly, it tries to emit `AddContent` if possible, and `AddMarkupContent` otherwise:

https://github.com/dotnet/razor/blob/dc75ec6b22ec555f649ad26060eebf40aa9077bb/src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentMarkupEncodingPass.cs#L38-L46

```cs
// Markup content in components are rendered in one of the following two ways,
// AddContent - we encode it when used with prerendering and inserted into the DOM in a safe way (low perf impact)
// AddMarkupContent - renders the content directly as markup (high perf impact)
// Because of this, we want to use AddContent as much as possible.
//
// We want to use AddMarkupContent to avoid aggressive encoding during prerendering.
// Specifically, when one of the following characters are in the content,
// 1. New lines (\r, \n), tabs (\t), angle brackets (<, >) - so they get rendered as actual new lines, tabs, brackets instead of &#xA;
// 2. Any character outside the ASCII range
```

1. If there are newlines, it cannot use `AddContent` because those newlines would be encoded; so it uses `AddMarkupContent` - with the HTML passed in verbatim as it appears in the razor file (if there was anything to encode, the user would have to encode it in the razor already);
2. otherwise, it *decodes* HTML entities (e.g., if user had written `&lt;` in the razor file, it's decoded into `<`) and uses `AddContent`.

My take:

1. I'm not sure if this optimization pass makes sense in the first place. Do we want to prefer `AddContent` over `AddMarkupContent` for perf reasons?
2. `AddContent` seems to be encoding single quotes `'` - is that necessary in HTML? If yes, maybe it could behave differently inside `<script>` tags?
3. I think we could disable this optimization pass in the razor compiler for `<script>` tags, so effectively only `AddMarkupContent` would be used for anything the user literally writes as HTML in their razor files.

**Comment by SteveSandersonMS** [comment]
> I'm not sure if this optimization pass makes sense in the first place. Do we want to prefer AddContent over AddMarkupContent for perf reasons?

I think it does make sense to do this optimization. What might be unclear from the discussion is that we have a single compilation output that has to work well in two very different rendering modes: interactive and SSR (aka prerendering).

I'll write up an overview of this here, even though I appreciate @jjonescz and @MackinnonBuck probably already know it, just to check if any of us including me are misunderstanding parts of it, and to give us something to refer back to if needed.

### The state today

 * Interactive rendering:
    * `AddMarkupContent` is a performance optimization when the source code contains a block of fixed literal HTML markup containing nontrivial content (at least one HTML element, not just a text literal). It's a performance optimization in the sense that it's **faster for the interactive renderer than building that content element-by-element and attribute-by-attribute**. It lets the interactive renderer just use the native DOM parser in the browser to attach it all in a single step. This is only safe because we know it comes purely from fixed literal markup - we cannot use it if there's any code expression (otherwise it would be an XSS risk or just corrupt the DOM if the code evaluates to something containing HTML tags or elements).
    * When there is a code expression, we **must** use a combination of `OpenElement`/`AddAttribute`/`AddContent`/etc. `AddContent` causes us to insert the value simply by setting `elem.textContent = value`. This is XSS safe and very performant. It's how something like `<div>@value</div>` writes its value interactively.
    * Which is faster? `AddMarkupContent` is faster if we're dealing with a literal block containing multiple HTML elements etc. (compared with having lots of `OpenElement`/`AddAttribute`/`AddContent`/etc. calls), whereas `AddContent` is faster than `AddMarkupContent` if we're only inserting a single piece of text content (plus it avoids XSS issues).
 * SSR (aka prerendering):
   * `AddMarkupContent` just writes the output to the response verbatim, no HTML encoding involved. This produces the same effective behavior as for interactive rendering.
   * `AddContent` writes the output to the response with HTML encoding. This produces the same effective behavior as for interactive rendering. Unfortunately this means that if the text contains a newline/etc, it will be emitted as an HTML entity even though that's unnecessary and makes the HTML less readable.
   * Which is faster? `AddMarkupContent` because it skips HTML encoding, but it can only be used if there's no code expression.

In the case of text content that contains no code expressions, it's technically legal to use either `AddContent(HtmlDecode(sourceCode))` or `AddMarkupContent(sourceCode)`, since both will produce the same effect in both interactive rendering and SSR. So which should we use in that case?

 * For perf, the interactive renderer would prefer `AddContent`
 * For perf, the static renderer would prefer `AddMarkupContent`

So we could go either way. However if the text content contains newlines, it's kind of annoying for SSR to emit them as HTML entities, so for the sake of making "view source" more readable, we've chosen to use `AddMarkupContent`.

### Implications for `<script>`

 * For interactive rendering:
   * It never makes sense to build the content so of a `<script>` block through multiple calls to `AddContent` or `AddMarkupContent`, as that would never work because there is no DOM inside a `<script>` block. There is only a single `.textContent` that can be assigned. So if there was anything beside a single code expression (`<script>@expression</script>`) or literal text (`<script>mySourceCode...</script>`) then it will create a runtime error.
 * For SSR:
   * It would actually work to have multiple calls to `AddContent` and/or `AddMarkupContent`, because we're just building a response string by concatenating stuff.
   * `AddMarkupContent` would need to be used for all literal content from the source code, regardless of whether it contains newlines or whatever else. That's because we can't HTML-encode any of it (it would turn `alert('hi')` into `alert(&apos;hi&apos;)` which is an error).
   * `AddContent` would need to be used for any code expressions, otherwise it's an XSS vector. **And critically, the runtime SSR renderer would have to use JavaScriptStringEncoding not HTML encoding** (I don't know if it already does).

Does that sound right? In the past we didn't really need to deal with this because, as mentioned, it didn't work in interactive rendering anyway. But now we have more SSR scenarios we can make it work for SSR at least if we want (even though it will still fail for interactive rendering at runtime).

### Questions

> I'm not sure if this optimization pass makes sense in the first place. Do we want to prefer AddContent over AddMarkupContent for perf reasons?

Yes, we do want to prefer AddContent over AddMarkupContent for perf for the interactive renderer as described above. And in the case of code expressions, `AddContent` is the only legal choice.

> AddContent seems to be encoding single quotes ' - is that necessary in HTML? If yes, maybe it could behave differently inside <script> tags?

It's necessary to encode `'` when writing out HTML attributes. AFAIK yes it could work differently inside `<script>` tags, but I think we could use `AddMarkupContent` for all literal sources (not code expressions) inside `<script>` and then no encoding would be involved anyway.

> Alternatively, we could disable this optimization pass in the razor compiler for <script> tags, so effectively only AddMarkupContent would be used for anything the user literally writes as HTML in their razor files.

Agreed, but only inside `<script>` blocks.

> Probably simplest: we can update the optimization pass to fallback to AddMarkupContent if the string contains single quotes (similarly to how it handles newlines, tabs, angle brackets, and characters outside ASCII today). Are there other characters to be handled like that?

I'd prefer if we don't do that, as it will harm perf for interactive rendering for anything containing apostropes, e.g., `<h1>It's a nice day, @name!</h1>`.
